chall05/mzomeno-.c

```diff
@@ -0,0 +1,51 @@
+#include <stdio.h>
+
+short ft_ie_except_after_c(char *str)
+{
+  /*
+  * First, we make the first letter lowercase by applying the following mask:
+  * 32 = 00100000
+  * Thus, if bit 5 was 0 (uppercase letters always have their bit five put to 0),
+  * we would set it to 1 (so the resultant byte would be the correspondant
+  * lowercase letter). Else, the byte would remain the same.
+  *
+  * Note that uppercase letters in the middle of the word are not taken into account
+  */
+  /**
+   ** I wanted to put the following line of code into the function:
+   ** *str |= 32;
+   ** But it doesn’t compile. It throws a segfault error, I haven’t figured out why
+   **/
+  while (*str)
+    { 
+      /*
+       * Now, we check out the "cie" cases
+       */
+      if (*str == 'c' && *(str + 1) == 'i' && *(str + 2) == 'e')
+      return (0);
+      /*
+       * Then, we check out the "ei" cases which aren’t preceeded by ’c’
+       */
+      if (*str == 'e' && *(str + 1) == 'i' && *(str - 1) != 'c')
```

Comment:
Checking for `index - 1` is usually very dangerous, you always need to check your bounds.

In this case, you go out of bounds for `str = "ei"`, the program doesn't crash, but it's still very very dangerous.

Reply:
I heard about it but I still don't understand why is it so problematic to read memory outside the string. I can research it on my own, anyways.

Reply:
If you read out of bounds you will most likely read adjacent variables, or garbage memory. It causes undefined behavior and if you were to interact with that memory it could cause all kind of problems.